spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/MqttPahoMessageDrivenChannelAdapter.java

```diff
@@ -263,22 +272,38 @@ private synchronized void connectAndSubscribe() throws MqttException { // NOSONA
 		}
 		Assert.state(getUrl() != null || connectionOptions.getServerURIs() != null,
 				"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
-		this.client = this.clientFactory.getClientInstance(getUrl(), getClientId());
-		this.client.setCallback(this);
-		if (this.client instanceof MqttClient) {
-			((MqttClient) this.client).setTimeToWait(getCompletionTimeout());
+
+		IMqttAsyncClient clientInstance;
+
+		if (getClientManager() == null) {
+			this.client = this.clientFactory.getAsyncClientInstance(getUrl(), getClientId());
+			this.client.setCallback(this);
+			clientInstance = this.client;
+		}
+		else {
+			clientInstance = getClientManager().getClient();
 		}
 
 		this.topicLock.lock();
 		String[] topics = getTopic();
 		ApplicationEventPublisher applicationEventPublisher = getApplicationEventPublisher();
 		try {
-			this.client.connect(connectionOptions);
-			this.client.setManualAcks(isManualAcks());
+			long completionTimeout = getCompletionTimeout();
+			if (!clientInstance.isConnected()) {
+				clientInstance.connect(connectionOptions).waitForCompletion(completionTimeout);
+			}
+			clientInstance.setManualAcks(isManualAcks());
```

Comment:
This line is also a potential issue if multiple adapters with different settings reuse the client

Reply:
Right. This is one which has to be an option on the `ClientManager` by itself.
Remember: if we provide an external client, then no any local mutation to it.
That's why it feels like an internal `ClientManager` abstraction fits more and more to our expectations.

Reply:
So, I have moved manualAcks to ClienManager as a property (which probably may be injected from XML configuration)

Reply:
"Moved"?

I thought we agreed to keep the current behavior...
Therefore `setManualAck()` still must be present on the `AbstractMqttMessageDrivenChannelAdapter` for cases when external client is not provided.

Reply:
Sorry, not completely moved. It is still used for local clients of course.